### ISSUE TITLE
Need help with Zoho Integration

### ISSUE THREAD

**Issue by robzor** [verbb/formie#410]
**Description**

Hi there,

I'm unsure if this is an issue on my end or not but this is the first time I've had to integrate with Zoho. We've got a client who wants us to map some 2 forms to leads in Zoho and the field mapping has gone fine using your lovely settings panel. However, i am getting a Craft console error `Triggering form “essenzaZoho” integration.` and no lead is being generated:

The error:

`API error: “Missing return “leadId” {"data":[{"code":"INVALID_DATA","details":{"expected_data_type":"jsonarray","api_name":"Destination"},"message":"invalid data","status":"error"}]}”`

I've looked into this and realised that the Zoho 'Destination' field is set up as a dropdown field in Zoho, and in Formie I am using an Entries field.

Do you have any ideas?

![CleanShot 2021-03-25 at 3 42 56](https://user-images.githubusercontent.com/704032/112501130-e1e2ce80-8d80-11eb-9e12-2be59073512b.png)

**Steps to reproduce**
1. Create a form and assocate with Zoho Lead
2. Submit the form on the front end

### Form settings
Please pick from the options below.

- Multi-page form: No
- Submission Method: Page Reload
- Client-side Validation: Yes 
- Custom Form Templates: No

**Additional info**
- Plugin version: 1.3.25
- Craft version: 3.6.11.1
- Multi-site: No

**Additional context**

**Comment by robzor** [comment]
An additional thing to add here is that the Phone field is also causing this error:

`{"expected_data_type":"phone","api_name":"Phone"}`

This is with a Phone Number type field in Formie and a Phone type field in Zoho. In the Formie integration mapping, I have chosen 'Phone number' rather than any of the subtypes.

Full error:

`API error: “Missing return “leadId” {"data":[{"code":"INVALID_DATA","details":{"expected_data_type":"phone","api_name":"Phone"},"message":"invalid data","status":"error"}]}”`

![CleanShot 2021-03-25 at 3 59 22](https://user-images.githubusercontent.com/704032/112503845-2c654a80-8d83-11eb-9110-98f4f66e25ee.png)

**Comment by engram-design** [comment]
So for the first error, it seems to be an issue with the `Destination` field (in Zoho), when creating a Lead, and it expects an array. Can you let me know what type of field you're mapping to that value (in Formie)?

Similarly for Phone - I imagine that's just a phone number field (in Formie)?

**Comment by robzor** [comment]
Hi there, the Destination field is an entries field in Formie and the Phone one is the Phone Number field in Formie. See above screenshot for the Phone mapping

Thanks!

**Comment by engram-design** [comment]
Does your phone field have a country dropdown enabled? I can't seem to replicate it not liking the phone number with or without it enabled.

What sort of field type is the `Destination` field in Zoho?

**Comment by robzor** [comment]
Phone number in Formie has the country drop-down enabled, 'Destination' is a drop-down field in Zoho

**Comment by robzor** [comment]
Will an entries field by default just pass through a string of the entry title? I might be able to swing the client to make another field that is just plain text.

The phone number one might be a harder sell though

**Comment by engram-design** [comment]
What Phone Number value are you testing with? I'm wondering if we might need to massage it a little to play nice with Zoho's own validation. However, Formie does send phone numbers in an international format like `+61 3 6482 XXXX`, which clearly depends on the country, but I wouldn't have thought it be an issue with Zoho.

With entries, it depends on the destination field. If it's a single-value field (like a text field, or dropdown field), it should return just the title of the entry. For multiple-value fields (checkboxes) it'll return an array of entry titles.

Interestingly, it looks like a Pick List (what Zoho calls a dropdown) will actually create an option from a value you provide it. So even if "My Test Entry" doesn't exist in your Pick List field in Zoho, it'll create that option for you.

Here's also the result of a lead with `Craft 3.5 Supports GraphQL Mutations` being the name of an entry I'm testing with, mapped to a Pick List and Text field.

![image](https://user-images.githubusercontent.com/1221575/112559829-e7462480-8e25-11eb-8d5e-321dd74873ed.png)

**Comment by robzor** [comment]
Hi, sorry for the delay in getting back to you, I'm in the UK so yay timezones!

How interesting, maybe I should try creating some new fields in Zoho and test that way. The phone number was a +44 one. It's slightly annoying that the Zoho API just says that it expects a JSON array (for the drop-down) and a 'phone' rather than any more in-depth detail!

Is there any more info I could give you or any pointers you could give me to try and debug this?

Thanks for all your help so far

Rob

![cleanshot_2021-03-25_at_3.55.34.png](https://user-images.githubusercontent.com/704032/112594031-36795e80-8e00-11eb-95f9-22c1739ff012.png)

**Comment by engram-design** [comment]
No problem regarding delay! Hmmm, so I've pushed some additional logging to the latest `craft-3` branch which might help, in that it'll show what the payload you're sending to Zoho. That might be helpful in my reproduction of the issue.

If you're up for it, you can install that version and re-run the queue job to get that information. Change your `verbb/formie` requirement in `composer.json` to:

```
"require": {
  "verbb/formie": "dev-craft-3 as 1.3.25",
  "...": "..."
}
```

Then run `composer update`. Then, if you'd like to post back the error response I can see what I can do about replicating it.

**Comment by robzor** [comment]
Ah that is very useful. 

So, for the 'Phone Number' field, I've looked at the form entries that the client was using, and it appears that they didn't fill out the phone number field at all (it's not required). With your extra logging we get this:

`API error: “Missing return “leadId” {"data":[{"code":"INVALID_DATA","details":{"expected_data_type":"phone","api_name":"Phone"},"message":"invalid data","status":"error"}]}. Sent payload {"data":[{"First_Name":"Test","Last_Name":"Person","Email":"rob@highpass.co.uk","Phone":"() Not provided.","Anything_else":"This is an example test entry","Wishlist_Items":"","Referral_Page":"Property: Villa/Chalet - Villa Cortona","How_many_people_in_your_group":"3-4 People"}]}”`

If I then create a new submission with a phone number:

`Error | API error: “Missing return “leadId” {"data":[{"code":"INVALID_DATA","details":{"expected_data_type":"integer","maximum_length":9,"api_name":"How_many_people_in_your_group"},"message":"invalid data","status":"error"}]}. Sent payload {"data":[{"First_Name":"Test","Last_Name":"Person","Email":"rob@highpass.co.uk","Phone":"+44 7779 330558","Anything_else":"This is a test submission to test Zoho integration.","Wishlist_Items":"","Referral_Page":"Property: Villa/Chalet - Villa Cortona","How_many_people_in_your_group":"1-2 People"}]}”`

So the second one is another new error, but it seems that the field they want us to map to in Zoho only supports 9 character integers so that's something we can change in Zoho.

Just going to test the other form (With the Destination field). Stay tuned!

**Comment by robzor** [comment]
And the Destination field error is:

`API error: “Missing return “leadId” {"data":[{"code":"INVALID_DATA","details":{"expected_data_type":"jsonarray","api_name":"Destination"},"message":"invalid data","status":"error"}]}. Sent payload {"data":[{"First_Name":"Test","Last_Name":"Person","Phone":"+44 7779 330558","Anything_else":"Test submission for Zoho integration testing","Destination":"Florence","Wishlist_Items":"","Referral_Page":"","Experience_of_Interest":"Art & Culture","When_are_you_travelling":"Within next 2 - 3 months"}]}”`

**Comment by robzor** [comment]
Not sure if this is helpful but I've managed to find the edit layout bit in Zoho that some of these fields are mapped to:

![CleanShot 2021-03-26 at 7 39 03](https://user-images.githubusercontent.com/704032/112598374-6461a180-8e06-11eb-84d1-7498424e60c6.png)

Looks like that Destination field is a multi-select?

**Comment by robzor** [comment]
Ok so this is what I've done:
1) Change Zoho to make the destination field a single line
2) Change the 'Pax' field to a single line not a number

This has fixed both of those issues and they are now populating in Zoho as Leads.

So then the only outstanding thing is a blank phone number not working?

**Comment by robzor** [comment]
Ahh!  

So if the phone number is blank, it's sending `"() Not provided."` as a string to Zoho and Zoho is saying this is an invalid value.

**Comment by engram-design** [comment]
Ah, good investigation here. We should probably take care of a multi-select case, so I'll look into that.

However, the phone field issue is another issue, and that should be fixed (just run `composer update` again).

**Comment by robzor** [comment]
Ok, I'll test that now :) Thank you

**Comment by robzor** [comment]
Works fine on this end, excellent! When do you think you will have a chance to make it live?

Really appreciate all your help with this :)

**Comment by engram-design** [comment]
I'll just be doing a few more things over the weekend, so expect a release early next week. Thanks for sticking with this one!

**Comment by robzor** [comment]
Excellent, I've let the client know they need to add a Phone number for now and that a fix is on it's way :)

Thank you again, have a good weekend!

**Comment by engram-design** [comment]
Fixed in [1.3.26](https://github.com/verbb/formie/releases/tag/1.3.26)
